packages/vesting-base/src/contract.rs

```diff
@@ -0,0 +1,535 @@
+            return true;
+        }
+        false
+    }
+
+    /// Receives a message of type [`Cw20ReceiveMsg`] and processes it depending on the received template.
+    ///
+    /// * **cw20_msg** CW20 message to process.
+    fn receive_cw20(
+        &self,
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        cw20_msg: Cw20ReceiveMsg,
+    ) -> Result<Response, ContractError> {
+        let config = self.config.load(deps.storage)?;
+
+        // Permission check
+        if !self.is_sender_whitelisted(
+            deps.as_ref(),
+            &config,
+            &deps.api.addr_validate(&cw20_msg.sender)?,
+        ) || token_asset_info(info.sender) != config.vesting_token
+        {
+            return Err(ContractError::Unauthorized {});
+        }
+
+        match from_binary(&cw20_msg.msg)? {
+            Cw20HookMsg::RegisterVestingAccounts { vesting_accounts } => self
+                .register_vesting_accounts(
+                    deps,
+                    vesting_accounts,
+                    cw20_msg.amount,
+                    env.block.height,
+                ),
+        }
+    }
+
+    /// Adds new vesting managers, which have a permission to add/remove vesing schedule
+    ///
+    /// * **managers** list of accounts to be added to the whitelist.
+    pub fn add_vesting_managers(
+        &self,
+        deps: DepsMut,
+        _env: Env,
+        info: MessageInfo,
+        managers: Vec<String>,
+    ) -> Result<Response, ContractError> {
+        let config = self.config.load(deps.storage)?;
+        if info.sender != config.owner {
+            return Err(ContractError::Unauthorized {});
+        }
+        let mut attrs: Vec<Attribute> = vec![];
+        for m in managers {
+            let ma = deps.api.addr_validate(&m)?;
+            self.vesting_managers.save(deps.storage, &ma, &())?;
+            attrs.push(attr("vesting_manager", &m))
+        }
+        Ok(Response::new()
+            .add_attribute("action", "add_vesting_managers")
+            .add_attributes(attrs))
+    }
+
+    /// Removes new vesting managers from the whiltelist
+    ///
+    /// * **managers** list of accounts to be removed from the whitelist.
+    pub fn remove_vesting_managers(
+        &self,
+        deps: DepsMut,
+        _env: Env,
+        info: MessageInfo,
+        managers: Vec<String>,
+    ) -> Result<Response, ContractError> {
+        let config = self.config.load(deps.storage)?;
+        if info.sender != config.owner {
+            return Err(ContractError::Unauthorized {});
+        }
+        let mut attrs: Vec<Attribute> = vec![];
+        for m in managers {
+            let ma = deps.api.addr_validate(&m)?;
+            self.vesting_managers.remove(deps.storage, &ma);
+            attrs.push(attr("vesting_manager", &m))
+        }
+        Ok(Response::new()
+            .add_attribute("action", "remove_vesting_managers")
+            .add_attributes(attrs))
+    }
```

Comment:
If manager didn't exist in a first place, should we still print to attributes about its deletion?

Reply:
agree, we don't check if a manager from the given list actually exists in the vesting_managers list. so if we passed a list of addresses that aren't managers, the result attributes will show us a `vesting_manager` key by each passed address regardless of they weren't managers actually